### PR TITLE
Implement meta tx concept

### PR DESCRIPTION
**Pull request by jsidorenko**
This PR adds implements the meta tx concept.
A user could pre-sign a transaction and pass it to the relayer that would actually submit it by paying the fee.
A relayer could be whitelisted within the call object to prevent the front-running.

### CODE REVIEW

File: frame/utility/src/tests.rs [hunk 1]
```diff
@@ -638,3 +668,92 @@ fn force_batch_works() {
...
+			forwarded_call.clone(),
+			signature.clone()
+		));
+
+		// validate event
+		System::assert_last_event(
+			utility::Event::CallForwarded {
+				forwarder: 1,
+				new_origin: signer_id,
+				call_hash: blake2_256(&Encode::encode(&call)),
```
Comment (shawntabrizi) [inline frame/utility/src/tests.rs:703]: how does this work? how do you successfully sign a message with account "1"?

Seems you need to use real accountid32s here

File: frame/utility/src/lib.rs [hunk 2]
```diff
@@ -459,6 +483,63 @@ pub mod pallet {
...
 		}
+
+		/// Forwards a pre-signed function call.
+		///
+		/// # <weight>
+		/// - O(1).
+		/// - Limited storage reads.
+		/// - One DB write (event).
+		/// - Weight of derivative `call` execution + T::WeightInfo::dispatch_as().
+		/// # </weight>
```
Comment (shawntabrizi) [inline frame/utility/src/lib.rs:494]: delete this stuff, and add a bit more documentation about the parameters of the call below and when / how to use this function

File: frame/utility/src/tests.rs (hunk 1 above)
Comment (jsidorenko) [reply frame/utility/src/tests.rs:703]: `forwarder` has the AccountId type, it's a regular account that will submit the tx and pay the fee

File: frame/utility/src/lib.rs (hunk 2 above)
Comment (jsidorenko) [reply frame/utility/src/lib.rs]: done

File: frame/utility/src/lib.rs [hunk 3]
```diff
@@ -459,6 +483,66 @@ pub mod pallet {
...
 			Ok(Some(base_weight + weight).into())
 		}
+
+		/// Forwards a pre-signed function call.
+		///
+		/// This method allows to dispatch the pre-signed call by relayer on user's behalf
+		///
+		/// - `forwarded_call`: A struct that consists:
+		/// 	- `call`: the call to be forwarded
+		/// 	- `nonce`: a nonce of the account that signed the call
```
Comment (shawntabrizi) [inline frame/utility/src/lib.rs:493]:
```suggestion
		/// 	- `forwarded_call`: A struct that consists:
		/// 	- `call`: the call to be forwarded
		/// 	- `nonce`: a nonce of the account that signed the call
```

**Review by shawntabrizi — APPROVED** [review]

File: frame/utility/src/lib.rs [hunk 4]
```diff
@@ -478,3 +562,25 @@ impl<T: Config> Pallet<T> {
 			.expect("infinite length input; no invalid inputs for type; qed")
 	}
 }
+
+#[derive(Encode, Decode, Clone, PartialEq, RuntimeDebug, TypeInfo)]
+#[codec(dumb_trait_bound)]
```
Comment (bkchr) [inline frame/utility/src/lib.rs:567]:
```suggestion
```

Also this type misses docs.

File: frame/utility/src/lib.rs [hunk 5]
```diff
@@ -478,3 +562,25 @@ impl<T: Config> Pallet<T> {
...
+pub struct ForwardedCall<Nonce, BlockNumber, Call, AccountId> {
+	pub nonce: Nonce,
```
Comment (bkchr) [inline frame/utility/src/lib.rs:569]: Docs missing for each of these fields.

File: frame/utility/src/lib.rs [hunk 6]
```diff
@@ -478,3 +562,25 @@ impl<T: Config> Pallet<T> {
...
+	pub deadline: BlockNumber,
+	pub call: Box<Call>,
+	pub signer: MultiSigner,
```
Comment (bkchr) [inline frame/utility/src/lib.rs:572]: Why is the signer not generic?

File: frame/utility/src/lib.rs [hunk 7]
```diff
@@ -459,6 +483,66 @@ pub mod pallet {
...
+			}
+
+			// validate deadline
+			let now = frame_system::Pallet::<T>::block_number();
+			ensure!(deadline >= now, Error::<T>::AuthorizationExpired);
+
+			// convert AccountId32 to AccountId
+			let who = who.into_account();
+			let who = T::AccountId::decode(&mut AccountId32::as_ref(&who))
+				.map_err(|_| Error::<T>::ErrorConvertingToAccountId)?;
```
Comment (bkchr) [inline frame/utility/src/lib.rs:528]: Why is all of this not generic?

File: frame/utility/src/lib.rs [hunk 8]
```diff
@@ -459,6 +483,66 @@ pub mod pallet {
...
+
+			// validate nonce
+			let account = Account::<T>::get(&who);
+			if nonce != account.nonce {
+				return Err(Error::<T>::InvalidNonce.into())
+			}
```
Comment (bkchr) [inline frame/utility/src/lib.rs:534]: Here you check the nonce, but you never increase it?

File: frame/utility/src/lib.rs [hunk 9]
```diff
@@ -459,6 +483,66 @@ pub mod pallet {
...
+
+			// emit event and dispatch call
+			let call_hash = blake2_256(&Encode::encode(&call));
```
Comment (bkchr) [inline frame/utility/src/lib.rs:537]:
```suggestion
			let call_hash = T::Hash::hash(&Encode::encode(&call));
```

Why not the block hash?

Comment (bkchr) [reply frame/utility/src/lib.rs:537]: And you also encode the call two times. Once here and then inside the `verify` function.

**Review by bkchr — COMMENTED** [review]
I don't really get this merge of generic/non-generic code.

File: frame/utility/src/lib.rs (hunk 6 above)
Comment (shawntabrizi) [reply frame/utility/src/lib.rs:571]: Where do we provide a generic signer? AFAIK MultiSigner was the closest we got?

Comment (bkchr) [reply frame/utility/src/lib.rs:571]:
```
trait Config {
    type Signature: Verify<Signer = Self::Signer>;
    type Signer: IdentifyAccount<AccountId = <Self as frame_system::Config>::AccountId;
}
```

File: frame/utility/src/lib.rs (hunk 8 above)
Comment (shawntabrizi) [reply frame/utility/src/lib.rs:534]: needs a test too

Comment (ggwpez) [reply frame/utility/src/lib.rs:534]: The current logic invalidates the TX as soon as the account submits another TX.  
Could be done by design.

File: frame/utility/src/lib.rs (hunk 6 above)
Comment (ggwpez) [reply frame/utility/src/lib.rs:571]: Basically how I explained it on [SO](https://substrate.stackexchange.com/questions/305/how-to-verify-a-signature-in-a-pallet).  
Full pallet example [here](https://github.com/perun-network/perun-polkadot-pallet/blob/b5a1897cdf5c7755cddfb2b1841331acab349aba/src/lib.rs#L98) and [here](https://github.com/perun-network/perun-polkadot-pallet/blob/b5a1897cdf5c7755cddfb2b1841331acab349aba/src/types.rs#L177).  
This manual sig verification stuff is a bit tricky.

File: frame/utility/src/benchmarking.rs [hunk 10]
```diff
@@ -86,5 +112,33 @@ benchmarks! {
 		assert_last_event::<T>(Event::BatchCompleted.into())
 	}
 
+	submit_call_on_behalf_of {
+		let caller: T::AccountId = account("caller", SEED, SEED);
+		let call = Box::new(frame_system::Call::remark { remark: vec![] }.into());
```
Comment (ggwpez) [inline frame/utility/src/benchmarking.rs:117]: I think to get the maximum cost for the `Encode::encode` function we need to use a deeply nested call here.  
Or a rather long one, not sure what is more expensive.

File: frame/utility/src/lib.rs (hunk 8 above)
Comment (bkchr) [reply frame/utility/src/lib.rs:534]: We should probably provide a custom `SignedExtension` that returns a `provides` which contains the `account.nonce`. So, that the `account` could send "normal" transactions that depend on this "wrapped" transaction.

**Comment by xlc** [comment]
We have already implemented https://github.com/AcalaNetwork/Acala/pull/2029

File: frame/utility/src/lib.rs (hunk 9 above)
Comment (jsidorenko) [reply frame/utility/src/lib.rs:537]: Inside the `verify` function I encode the whole `ForwardedCall` struct, while here it encodes only the inner `call` param

**Pull request closed without merging**